File: config/src/main/java/com/typesafe/config/impl/ResolveStatus.java
/**
 *   Copyright (C) 2011-2012 Typesafe Inc. <http://typesafe.com>
 */
package com.twitter_typesafe.config.impl;

import java.util.Collection;

/**
 * Status of substitution resolution.
 */
enum ResolveStatus {
    UNRESOLVED, RESOLVED;

    final static ResolveStatus fromValues(
            Collection<? extends AbstractConfigValue> values, Collection<ConfigConditional> conditionals) {
        for (AbstractConfigValue v : values) {
            if (v.resolveStatus() == ResolveStatus.UNRESOLVED)
                return ResolveStatus.UNRESOLVED;
        }
        if (conditionals.size() > 0)
            return ResolveStatus.UNRESOLVED;
        return ResolveStatus.RESOLVED;
    }

    final static ResolveStatus fromBoolean(boolean resolved) {
        return resolved ? ResolveStatus.RESOLVED : ResolveStatus.UNRESOLVED;
    }
}

File: config/src/main/scala/com/typesafe/config/impl/ResolveStatus.scala
package com.typesafe.config.impl

import java.{ util => ju }

/**
 * Status of substitution resolution.
 */
class ResolveStatus private (name: String, ordinal: Int)
    extends Enum[ResolveStatus](name, ordinal)

object ResolveStatus {

    final val UNRESOLVED = new ResolveStatus("UNRESOLVED", 0)
    final val RESOLVED = new ResolveStatus("RESOLVED", 1)

    private[this] val _values: Array[ResolveStatus] =
        Array(UNRESOLVED, RESOLVED)

    def values(): Array[ResolveStatus] = _values.clone()

    def valueOf(name: String): ResolveStatus =
        _values.find(_.name == name).getOrElse {
            throw new IllegalArgumentException(
                "No enum const ResolveStatus." + name)
        }

    def fromValues(values: ju.Collection[_ <: AbstractConfigValue]): ResolveStatus = {
        import scala.collection.JavaConverters._
        values.asScala.find(_.resolveStatus == ResolveStatus.UNRESOLVED) match {
            case Some(_) => ResolveStatus.UNRESOLVED
            case None => ResolveStatus.RESOLVED
        }
    }

    def fromBoolean(resolved: Boolean): ResolveStatus =
        if (resolved) ResolveStatus.RESOLVED else ResolveStatus.UNRESOLVED

}